### ISSUE TITLE
Regression in measuring edge_count in fuzzer_stats?

### ISSUE THREAD

**Issue by mhlakhani** [AFLplusplus/AFLplusplus#873]
Hi!

I'm in the process of upgrading our version of AFL++ from 3.00c to the latest commit in the stable branch; and I noticed an issue in the reporting of `edge_count` inside the `fuzzer_stats` file. I'm not sure if I am using AFL++ incorrectly, or if this is a bug, or if this is an intentional behaviour change.

We currently run AFL++ in parallel across a number of cores, and at the end of the run, we copy all the queue entries into a new directory and then run `afl-fuzz` again passing `-V 0` so we can compute the `bitmap_coverage` and `edge_count` observed from the collected queue entries across all the jobs.

With v3.00c, this would work fine, and we would get a number that was >= the numbers in the original `fuzzer_stats` files.
With the latest version, I'm very consistently getting `edge_count` = 0 across a number of targets.

Is this a regression, or should we be directly looking at the fuzzer_stats from the original run without trying to re-run it after combining all the queue entries?

**Comment by vanhauser-thc** [comment]
what is the exact command you are using with -V0 ?
and can you paste a fuzzer_stats please with edge_count 0 ?

**Comment by mhlakhani** [comment]
@vanhauser-thc:

We first run the following commands in parallel:

    AFL_EXPAND_HAVOC_NOW=1 AFL_FAST_CAL=1 AFL_I_DONT_CARE_ABOUT_MISSING_CRASHES=1 AFL_NO_AFFINITY=1 AFL_NO_UI=1 AFL_SKIP_CRASHES=1 /tmp/afl-fuzz -i /tmp/corpus/JO6fwfrSz3mrVEDAHVEfK6WpHgxbvlPAAAM -o /tmp/scratch/afl_sync -M fuzzer01 -s 0 -V 60 -t 60000+ -m none /tmp/harnesses/target_name -2147483647	
    AFL_EXPAND_HAVOC_NOW=1 AFL_FAST_CAL=1 AFL_I_DONT_CARE_ABOUT_MISSING_CRASHES=1 AFL_NO_AFFINITY=1 AFL_NO_UI=1 AFL_SKIP_CRASHES=1 /tmp/afl-fuzz -i /tmp/corpus/JO6fwfrSz3mrVEDAHVEfK6WpHgxbvlPAAAM -o /tmp/scratch/afl_sync -S fuzzer02 -s 0 -V 60 -t 60000+ -m none /tmp/harnesses/target_name -2147483647	

Once that's done, we copy all the entries from the queue directories back into the `/tmp/corpus` directory
And then run AFL again, with this `fuzzer_stats` file as below (this includes the command we run)

```
start_time        : 1618435360
last_update       : 1618435360
run_time          : 0
fuzzer_pid        : 3558342
cycles_done       : 0
cycles_wo_finds   : 0
execs_done        : 19755
execs_per_sec     : 19755000.00
execs_ps_last_min : 0.00
paths_total       : 6535
paths_favored     : 302
paths_found       : 0
paths_imported    : 0
max_depth         : 1
cur_path          : 71
pending_favs      : 302
pending_total     : 3250
variable_paths    : 4
stability         : 96.81%
bitmap_cvg        : 17.50%
unique_crashes    : 0
unique_hangs      : 0
last_path         : 0
last_crash        : 0
last_hang         : 0
execs_since_crash : 19755
exec_timeout      : 300
slowest_exec_ms   : 0
peak_rss_mb       : 0
cpu_affinity      : -1
edges_found       : 0
var_byte_count    : 49
havoc_expansion   : 1
testcache_size    : 55
testcache_count   : 1
testcache_evict   : 0
afl_banner        : fuzzer01
afl_version       : ++3.12c
target_mode       : persistent shmem_testcase deferred
command_line      : /tmp/afl-fuzz -i /tmp/corpus/GJO6fwfrSz3mrVEDAHVEfK6WpHgxbvlPAAAM -o /tmp/afl_sync_metrics -M fuzzer01 -s 0 -V 0 -t 60000+ -m none /tmp/harnesses/target_name -2147483647
```

**Comment by vanhauser-thc** [comment]
can you please retry with the current dev state?

also:
1. why are you running two instances in parallel? IMHO that just waits cpu cycles.
2. why are you not using afl-cmin? that is basically the purpose of the script :)

**Comment by mhlakhani** [comment]
> can you please retry with the current dev state?

Will do!

> why are you running two instances in parallel? IMHO that just waits cpu cycles.

one is with -M and one is -S. This was also from a testing setup, in our real production flows we're passing different flags (mutation strategies and the like) - which I hope would not be a waste of CPU?

> why are you not using afl-cmin? that is basically the purpose of the script :)

In our use case, we run the minimization step later, after also collecting seeds from runs on other machines/configs/fuzzers/etc. But we use this initial quick step to see if any of the single-machine runs found more edges, before running `afl-cmin`, as that can be expensive.

**Comment by mhlakhani** [comment]
This looks to be fixed in the current dev branch; and since I saw https://github.com/AFLplusplus/AFLplusplus/pull/871 is open I can wait for the new stable to be out.

Apologies for the thrash!

**Comment by vanhauser-thc** [comment]
@mhlakhani thank you for reporting, otherwise I would not have been aware.

what is actually your point of running the two instances with -V0 then? is it just to get % coverage information?
For this you can use afl-showmap for some months now which will do it much faster: `afl-showmap -C -i out -o /dev/null -- ./target -params @@` (the -C is the key)

**Comment by mhlakhani** [comment]
@vanhauser-thc  sorry for not being clearer, we first run the two instances with -V{LARGE_NUMBER} (I just did 60 for testing here) - but then the final run is just one instance with -V0, and that is indeed to get coverage information - we're looking for both bitmap_pct and edge_count.

I'll try `afl-showmap` now, thank you!
